crates/testing-framework/src/ephemeral/mod.rs

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright 2022 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+pub mod neiborhood;
+
+use self::neiborhood::{PeerSet, PeerWithNeighborhood};
+use crate::services::{services_to_call_service_closure, Service};
+
+use air_test_utils::{
+    test_runner::{create_avm, TestRunParameters, TestRunner},
+    RawAVMOutcome,
+};
+
+use std::{
+    borrow::Borrow,
+    cell::RefCell,
+    collections::{HashMap, HashSet},
+    hash::Hash,
+    rc::Rc,
+};
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct PeerId(String);
+
+impl PeerId {
+    pub fn new(peer_id: impl Into<String>) -> Self {
+        Self(peer_id.into())
+    }
+}
+impl From<String> for PeerId {
+    fn from(source: String) -> Self {
+        Self(source)
+    }
+}
+
+impl From<&str> for PeerId {
+    fn from(source: &str) -> Self {
+        Self(source.to_owned())
+    }
+}
+
+impl Borrow<str> for PeerId {
+    fn borrow(&self) -> &str {
+        &self.0
+    }
+}
+
+pub type Data = Vec<u8>;
+
+pub struct Peer {
+    peer_id: PeerId,
+    // We presume that only one particle is run over the network.
+    prev_data: Data,
+    runner: TestRunner,
+}
+
+impl Peer {
+    pub fn new(peer_id: impl Into<PeerId>, services: Rc<[Rc<dyn Service>]>) -> Self {
+        let peer_id = Into::into(peer_id);
+        let call_service = services_to_call_service_closure(services);
+        let runner = create_avm(call_service, &peer_id.0);
+
+        Self {
+            peer_id,
+            prev_data: vec![],
+            runner,
+        }
+    }
+
+    pub fn invoke(
+        &mut self,
+        air: impl Into<String>,
+        data: Data,
+        test_run_params: TestRunParameters,
+    ) -> Result<RawAVMOutcome, String> {
+        let mut prev_data = vec![];
+        std::mem::swap(&mut prev_data, &mut self.prev_data);
+        let res = self.runner.call(air, prev_data, data, test_run_params);
+        if let Ok(outcome) = &res {
+            self.prev_data = outcome.data.clone();
+        }
+        res
+    }
+}
+
+impl std::fmt::Debug for Peer {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Peer")
+            .field("peer_id", &self.peer_id)
+            .field("prev_data", &self.prev_data)
+            .field("services", &"...")
```

Comment:
It cannot be, unfortunately, as `.field` requies `&dyn Debug`, but not `&dyn Debug + ?Sized`:
```
error[E0277]: the size for values of type `str` cannot be known at compilation time
   --> crates/testing-framework/src/ephemeral/mod.rs:104:32
    |
104 |             .field("services", "...")
    |                                ^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `str`
    = note: required for the cast from `str` to the object type `dyn Debug`
help: consider borrowing the value, since `&str` can be coerced into `dyn Debug`
    |
104 |             .field("services", &"...")
    |                                +
```